File: scala/ninetynineproblem/src/main/scala/jp/co/teruuu/s99/P08.scala
package jp.co.teruuu.s99

object P08 {

  def compress[A](l: List[A]): List[A] = {
    def go(a: A, l: List[A]): List[A] = l match {
      case x :: xs if x == a => go(a, xs)
      case x :: xs => a :: go(x, xs)
      case Nil => a :: Nil
    }
    l match {
      case x :: xs => go(x, xs)
      case Nil => Nil
    }
  }
}

File: haskell/NinetyNineProbrem/src/P08.hs
module P08 (
    compress 
) where


-- compress :: Eq a => [a] -> [a]
-- compress = map head . group

myDropWhile :: (a -> Bool) -> [a] -> [a]
myDropWhile f [] = []
myDropWhile f (x:xs) = if f x 
    then myDropWhile f xs
    else (x:xs)

compress []     = []
compress (x:xs) = x : (compress $ myDropWhile (== x) xs)